crd-generator/api/src/main/java/io/fabric8/crd/generator/InternalSchemaSwaps.java

```diff
@@ -20,30 +20,69 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
 import java.util.StringJoiner;
 import java.util.stream.Collectors;
 
 public class InternalSchemaSwaps {
-  private final Map<Key, Value> swaps = new HashMap<>();
+  private Map<Key, Value> allSwaps = new HashMap<>();
+  private final Map<Key, Value> swaps;
+  private final Map<Key, Integer> swapDepths;
 
-  public void registerSwap(ClassRef definitionType, ClassRef originalType, String fieldName, ClassRef targetType) {
-    Value value = new Value(definitionType, originalType, fieldName, targetType);
+  public InternalSchemaSwaps() {
+    this(new HashMap<>(), new HashMap<>(), new HashMap<>());
+  }
+
+  private InternalSchemaSwaps(Map<Key, Value> swaps, Map<Key, Integer> swapDepths, Map<Key, Value> allSwaps) {
+    this.allSwaps = allSwaps;
+    this.swaps = swaps;
+    this.swapDepths = swapDepths;
+  }
+
+  public InternalSchemaSwaps branchDepths() {
+    InternalSchemaSwaps result = new InternalSchemaSwaps(this.swaps, new HashMap<>(), this.allSwaps);
+    result.swapDepths.putAll(this.swapDepths);
+    return result;
+  }
+
+  public InternalSchemaSwaps branchAnnotations() {
+    InternalSchemaSwaps result = new InternalSchemaSwaps(new HashMap<>(), this.swapDepths, this.allSwaps);
+    result.swaps.putAll(this.swaps);
+    return result;
+  }
+
+  public void registerSwap(ClassRef definitionType, ClassRef originalType, String fieldName, ClassRef targetType,
+      int depth) {
+    Value value = new Value(definitionType, originalType, fieldName, targetType, depth);
     swaps.put(new Key(originalType, fieldName), value);
+    if (allSwaps.put(new Key(originalType, fieldName), value) != null) {
+      // it's simplest for now to just disallow this
+      throw new IllegalArgumentException("Nested SchemaSwap: " + value);
+    }
   }
 
-  public Optional<ClassRef> lookupAndMark(ClassRef originalType, String name) {
-    Value value = swaps.get(new Key(originalType, name));
+  public ClassRef lookupAndMark(ClassRef originalType, String name, Runnable swapApplicableAction) {
```

Comment:
> I understand that, likely, this method is not used anywhere else outside of this codebase, still, it's public.

I just left things scoped as is, but I don't see a need to have InternalSchemaSwaps nor its methods greater than package scoped.  

Do you see some scenario where InternalSchemaSwaps is supposed to be used / extended by something else?

And/or should there be a consideration toward separating the annotations and truly public parts of the api to a separate dependency?  

> keep the original return type
> create an overload that respects the original signature

This depends on the answers to the above.

> instead of accepting a Runnable(which is overly open) we can pass in the visited Set<String>, in a first implementation we can simply run clear as your proposal already does, but we will leave the door open to a better refinement, if needed, in the future

Unfortunately this method is hiding a bit of complexity.  I don't really want to put much logic toward directly passing the visited as mentioned elsewhere that it's really the proper tracking of this situation.  

If anything I'd opt for returning more information for now to let the caller deal with the situation if you won't accept a runnable.

Reply:
> Do you see some scenario where InternalSchemaSwaps is supposed to be used / extended by something else?
And/or should there be a consideration toward separating the annotations and truly public parts of the api to a separate dependency?

That's a scary scenario ... but I cannot completely exclude it sadly, I think that e.g. the Quarkus Operator SDK is hooking into this mechanism cc. @metacosm @csviri

> If anything I'd opt for returning more information for now to let the caller deal with the situation if you won't accept a runnable.

I'm happy to go into this direction!

Reply:
> I'm happy to go into this direction!

That's been added - with the assumption that it's really only for use in this package.  Maybe it would be worth a follow task to discuss changing more of the apis visibility and/or a separate jar.